### PR TITLE
accurate emu timing

### PR DESCRIPTION
**Pull request by sprout42**
Made emulator timing flexible to support different methods of measuring time in the emulator. At the moment the emulator time is fixed at the tick-based time which should provide more accurate timing for the emulator execution.

The scaled emulation timing has the advantage that it removes the decision about when timers should expire from the primary thread of execution, because all time in the emulator is based on a fixed scaling factor off of the host OS's system time. The downside is that time will progress at a steady pace regardless of how efficient or inefficient the emulation of specific instructions or peripheral actions are. This can lead to timer events that should happen at deterministic points of code execution (like the decrementer exception or the ASIC watchdog expiring) happening much earlier than they are supposed to happen.

This PR implements tracking time by counting each instruction as a single tick. This isn't 100% accurate because there may be pipeline stalls depending on the sequence of loads, stores, or other instructions that take multiple cycles to execute. But without trying to emulate the full execution pipeline and caching this seemed like the way to make timing more accurate with the smallest impact to performance.

Unfortunately it looks like this change cuts performance almost in half given a brief run of the CM2350 firmware during start up:

python version: 
```
[ins] In [2]: import sys;sys.version
Out[2]: '3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]'
```
"accurate timing" (`EmuTimeCore`)
```
since start: 3773501 instructions in 62.097 secs: 60768.233 ops/sec
```

"scaled timing" (`ScaledEmuTimeCore`)
```
since start: 2659691 instructions in 24.875 secs: 106921.954 ops/sec
```

### CODE REVIEW

File: cm2350/e200_gdb.py [hunk 1]
```diff
@@ -206,7 +206,7 @@ def _serverSetHWBreak(self, addr):
         if addr in self._bpdata:
             raise Exception('Cannot add breakpoint that already exists @ 0x%x' % addr)
 
-        logger.debug('Adding new breakpoint: 0x%x' % addr)
+        logger.debug('Adding new breakpoint: -1x%x', addr)
```
Comment (ma77carpenter) [inline cm2350/e200_gdb.py:209]: ??

File: cm2350/e200z7.py [hunk 2]
```diff
@@ -108,115 +109,14 @@ def __init__(self, emu):
...
-import vivisect.impemu.platarch.ppc as vimp_ppc_emu
-#class PPC_e200z7(mmio.ComplexMemoryMap, eape.Ppc32EmbeddedEmulator, PpcEmulationTime):
-class PPC_e200z7(mmio.ComplexMemoryMap, vimp_ppc_emu.PpcWorkspaceEmulator, eape.Ppc32EmbeddedEmulator, PpcEmulationTime):
+import vivisect.impemu.emulator as vimp_emu
+#import vivisect.impemu.platarch.ppc as vimp_ppc_emu
+
+class PPC_e200z7(mmio.ComplexMemoryMap, vimp_emu.WorkspaceEmulator,
+                 eape.Ppc32EmbeddedEmulator, ppc_time.PpcEmuTime,
+                 #emutimers.ScaledEmuTimeCore, clocks.EmuClocks):
+                 emutimers.EmuTimeCore, clocks.EmuClocks):
```
Comment (ma77carpenter) [inline cm2350/e200z7.py:119]: Are these related?  Should we really need to subclass them all here?  
The EmuClocks seems like an odd name for a subclassable class.  I was surprised to see it here.

File: cm2350/e200z7.py [hunk 3]
```diff
@@ -544,8 +436,60 @@ def debug_client_detached(self):
         Handles the condition when a gdb client detaches. For now we will halt
         the emulator when this occurs.
         '''
-        self.queueException(intc_exc.GdbClientDetachEvent())
-        self.resume_exec()
+        # if va is already a breakpoint, don't do it!!
+        if va in self._bpdata:
+            return -1   # raise exception
+
+        brkbytes = self.arch.archGetBreakInstr()
```
Comment (ma77carpenter) [inline cm2350/e200z7.py:443]: Will this be the same for VLE and BOOKE?

File: cm2350/e200z7.py [hunk 4]
```diff
@@ -544,8 +436,60 @@ def debug_client_detached(self):
...
+        for tva in range(va, va+len(brkbytes)-1):
+            if tva in self._bpdata:
+                return -2   # raise exception
+
+        for tva in range(va - len(brkbytes)+1, va):
+            if tva in self._bpdata:
+                return -3   # raise exception
+
+        # store existing bytes
+        self._bpdata[va] = self.readMemory(va, len(brkbyes))
```
[the remark this answers is not in the document]
Comment (sprout42) [reply cm2350/e200z7.py:455]: It shouldn't be, this is obviously some sort of copy/paste error. it looks like this may be the result of a bad merge of some sort. I am tracking it down now.

[the remark this answers is not in the document]
Comment (sprout42) [reply cm2350/e200z7.py]: Fixed strange copy/paste error where some code from an older version of `cm2350/e200_gdb.py` got pasted over the `debug_client_detached()` function in `cm2350/e200z7.py`

File: cm2350/e200z7.py (hunk 3 above)
Comment (sprout42) [reply cm2350/e200z7.py]: The `archGetBreakInstr()` function is commented in the PowerPC architecture module. there are 3 different break instructions that could be returned depending on context and there is no good way to distinguish between which one gets returned `vivisect/envi/archs/ppc/__init__.py`. It would require knowledge of the address the breakpoint is being placed, and if the memory address is in a VLE code section the available space available for the break instruction to be inserted.
```
    def archGetBreakInstr(self):
        # BOOKE:        dnh     4C00018C
        # VLE 32bit:    e_dnh   7C0000C2
        # VLE 16bit:    se_dnh  000F
        return b'\x4C\x00\x01\x8C'
```

File: cm2350/e200z7.py (hunk 2 above)
Comment (sprout42) [reply cm2350/e200z7.py:119]: The `EmuClocks` module is a class that allows different peripherals to register a clock source that can then be retrieved in a generic way rather than requiring knowledge of which peripheral is providing the clock. This was added in [9091cdbf96ee7b8cc0e012a2fb4f2c9dfa2a34e4](https://github.com/Assured-Micropatching/CM2350-Emulator/commit/9091cdbf96ee7b8cc0e012a2fb4f2c9dfa2a34e4). A good example of this is in the `cm2350/peripherals/flexcan.py` file [line 1361](https://github.com/Assured-Micropatching/CM2350-Emulator/blame/54274f2be7debfe29dfbab9231ee3d42f32d69cd/cm2350/peripherals/flexcan.py#L1361)

File: cm2350/e200_gdb.py (hunk 1 above)
Comment (sprout42) [reply cm2350/e200_gdb.py:209]: Changing from the old python string formatting style `%` to using the logging arguments so that `addr` isn't evaluated unless debug is turned on.

Comment (ma77carpenter) [reply cm2350/e200_gdb.py:209]: I get that.  What's up with the "-1x"?  I'm assuming that's some automated editing thing that is wrong.  Please let me know if this is some new Python formatting thing.

Comment (ma77carpenter) [reply cm2350/e200_gdb.py:209]:
```suggestion
        logger.debug('Adding new breakpoint: 0x%x', addr)
```

**Comment by ma77carpenter** [comment]
Before merging, I did some side-by-side testing, and found anecdotal evidence that this PR actually runs faster-per-instruction, than the previous version.  I'm not sure what you were testing, @sprout42, but it's possible you had greatly optimized the version of the older code?

I checked out the repo twice, before and after, and rand them side-by-side (ie. at the same time, to make sure they were fighting for the same resources on this machine).

(I reran the test after merging, because the data scrolled off the screen.  For this, I cloned the repo twice, post-merge, and checked out the commit before merging in configurable_ecu_timing)
```
$ git status
On branch work_pre_PR4_merge
nothing to commit, working tree clean

$ ./ECU_emu.py 
Let's interact!
Python 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: ecui.run(silent=True)
0 ###### RESET  0x23ed0
  1 ====>> CALL 0x23ef8 -> 0x23c60 (0x4156100f, 0x1, 0x0, 0x4156700f, 0x4156900f)
...
        4 ====>> CALL 0xd7500 -> 0x129dac (0x0, 0x5bcb92, 0x0, 0xf0, 0x0)
        4 <<==== RET  0xd7504 <- 0x129dc0 (0x0)
      3 <<==== RET  0xd64c4 <- 0xd7518 (0x5bcb92)
    2 <<==== RET  0x403a0 <- 0xd6588 (0x5bcb92)
  1 ****** 0xe0a00 RFI    <- 0x40520
^C
since start: 72334781 instructions in 1397.147 secs: 51773.191 ops/sec
```


```
$ git status
On branch main
Your branch is up to date with 'origin/main'.

Changes not staged for commit:
<snip - code i'm working on, which should only slow things>
no changes added to commit (use "git add" and/or "git commit -a")

$ ./ECU_emu.py 
Let's interact!
Python 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: ecui.run(silent=True)
0 ###### RESET  0x23ed0
  1 ====>> CALL 0x23ef8 -> 0x23c60 (0x0, 0x1, 0x0, 0x0, 0x0)
...
    2 ====>> CALL 0x4009c -> 0xd8f10 (0x8000000, 0x10540000, 0x10000, 0x60098cdc, 0x4)
    2 <<==== RET  0x400a0 <- 0xd8f1c (0x8000000)
  1 ****** 0xe0a00 RFI    <- 0x40520
^C
since start: 79143597 instructions in 1396.830 secs: 56659.444 ops/sec
```
